### PR TITLE
Apply clang-tidy recommendations to more components

### LINKED ISSUE #874: gfortran-10.1.0 doesn't like z'00112233' used in drel.f90
Stop using hexadecimal initialization when not absolutely required.

### PR DESCRIPTION
**Pull request by KineticTheory**
### Background

* I am still trying to prepare Draco for a full CI check with clang-tidy.  See #869 for related changes.  Once these rules are working, we can expand the checks to include new patterns.
* Also fixes #874

### Description of changes

+ Use `std::array` instead of C-style arrays.
+ Use `nullptr` instead of `NULL`.
+ Replace `typedef` with `using`.
+ Use `auto` if the type is already provided by `static_cast`.
+ Use `std::move` for constructor arguments.
+ Use `= default` or `= delete` for trivial constructors.
+ Remove redundant `DLL_PUBLIC` decoration
+ Use `override` instead of `virtual` for overloaded member functions.
+ Use class member default initialization instead of constructor initialization.
+ Prune dead code blocks.
+ Make use of `emplace_back` for construct and append operations.

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis/Appveyor CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] Code reviewed/approved, sufficient DbC checks, testing, documentation
* WIP
  * [x] cdi_eospac tests are failing on CTS-1

### CODE REVIEW

**Comment by KineticTheory** [comment]
@attom @keadyk  If you have a few minutes, this PR could use a review.  The number of touched files is somewhat large.  Maybe you can pick a couple of directories to look at and leave the rest to someone else.

File: src/RTT_Format_Reader/RTT_Format_Reader.cc [hunk 1]
```diff
@@ -17,12 +17,12 @@ namespace rtt_RTT_Format_Reader {
...
  * \param RTT_File Mesh file name.
  */
-RTT_Format_Reader::RTT_Format_Reader(string const &RTT_File)
+RTT_Format_Reader::RTT_Format_Reader(string RTT_File)
     : header(), dims(), spNodeFlags(), spSideFlags(), spCellFlags(),
       spNodeDataIds(), spSideDataIds(), spCellDataIds(), spCellDefs(),
       spNodes(), spSides(), spCells(), spNodeData(), spSideData(),
       spCellData() {
-  readMesh(RTT_File);
+  readMesh(std::move(RTT_File));
```
Comment (attom) [inline src/RTT_Format_Reader/RTT_Format_Reader.cc:25]: I'm not up on the intricacies of moves and strings. This is a copy followed by a move?

File: src/cdi/CDI.cc [hunk 2]
```diff
@@ -885,7 +886,7 @@ CDI::SP_MultigroupOpacity CDI::mg(rtt_cdi::Model m, rtt_cdi::Reaction r) const {
  */
 CDI::SP_CPEloss CDI::eloss(rtt_cdi::CPModelAngleCutoff mAC, int32_t pz,
                            int32_t tz) const {
-  map_it entry = CPEloss_map.find(std::make_pair(pz, tz));
+  auto const entry = CPEloss_map.find(std::make_pair(pz, tz));
```
Comment (attom) [inline src/cdi/CDI.cc:889]: Why is auto preferred here?

File: src/RTT_Format_Reader/RTT_Format_Reader.cc (hunk 1 above)
Comment (KineticTheory) [reply src/RTT_Format_Reader/RTT_Format_Reader.cc:25]: That is correct. In this scenario, a copy must be made and the recommendation is to force the copy in the constructor function call reducing the number of steps required.  Gritty details are [here](https://releases.llvm.org/8.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/modernize-pass-by-value.html)

File: src/cdi_ipcress/IpcressDataTable.cc [hunk 3]
```diff
@@ -107,13 +107,6 @@ void IpcressDataTable::setIpcressDataTypeKey() const {
         ipcressDataTypeKey = "ragray";
         dataDescriptor = "Gray Rosseland Absorption";
         break;
-      // NOTE: THIS KEY DOES NOT ACTUALLY EVER EXIST, See LA-UR-01-5543
```
Comment (attom) [inline src/cdi_ipcress/IpcressDataTable.cc:110]: Do we want to keep a reference to this LA-UR somewhere? Something like `// To see what keys exist, see LA-UR-01-5543, "TITLE".`

File: src/cdi/CDI.cc (hunk 2 above)
Comment (KineticTheory) [reply src/cdi/CDI.cc:889]: The return type is fully defined by the type of `CPEloss_map`. Choosing the wrong type on the LHS will force a cast and possibly a deep object construction.

**Comment by attom** [comment]
Otherwise, LGTM

File: src/cdi_ipcress/IpcressDataTable.cc (hunk 3 above)
Comment (KineticTheory) [reply src/cdi_ipcress/IpcressDataTable.cc:110]: Good idea.  I'll add the full reference to the class header.

Comment (KineticTheory) [reply src/cdi_ipcress/IpcressDataTable.cc:110]: Fixed in 5a93fd0.

**Comment by keadyk** [comment]
I'm looking at this now, @KineticTheory !

File: src/norms/Norms_Base.cc [hunk 4]
```diff
@@ -15,14 +15,9 @@ using namespace rtt_norms;
...
-Norms_Base::Norms_Base(void)
+Norms_Base::Norms_Base()
     : d_sum_L1(-42.0), d_sum_L2(-42.0), d_Linf(-42.0), d_sum_weights(-42.0) {
-  reset();
-}
-
-//----------------------------------------------------------------------------//
-//! Destructor.
-Norms_Base::~Norms_Base() { /* empty */
+  Norms_Base::reset();
```
Comment (keadyk) [inline src/norms/Norms_Base.cc:20]: I'm curious about this -- why is this reset() call prefixed with the class?

File: src/meshReaders/Hex_Mesh_Reader.hh [hunk 5]
```diff
@@ -140,37 +140,32 @@ private:
  * <h3> Format Details </h3>
```
Comment (keadyk) [inline src/meshReaders/Hex_Mesh_Reader.hh:140]: It looks like most of the HTML (or whatever this is) was stripped out in the following lines, but not the <h3></h3> here -- why is that? Is this a clang-tidy oversight?

File: src/meshReaders/Hex_Mesh_Reader.cc [hunk 6]
```diff
@@ -217,23 +219,19 @@ Hex_Mesh_Reader::get_unique_element_types() const {
...
+ *   faces.
  */
 std::map<std::string, std::set<unsigned>>
 Hex_Mesh_Reader::get_element_sets() const {
   // Alternatively, the private data of the class could be changed so that the
-  // work done here is done in the constructor. This would be more efficient
-  // if this is going to be used repetively.
-  typedef std::map<std::string, std::set<unsigned>> resultT;
+  // work done here is done in the constructor. This would be more efficient if
+  // this is going to be used repetively.
```
Comment (keadyk) [inline src/meshReaders/Hex_Mesh_Reader.cc:233]: This might be outside scope, but "repetitively" is misspelled :)

**Comment by keadyk** [comment]
I have a bit of a hard time parsing the codecov/project results... is it saying this PR causes coverage of CDI to decrease? I'm not sure that makes sense to me.

File: src/norms/Norms_Base.cc (hunk 4 above)
Comment (KineticTheory) [reply src/norms/Norms_Base.cc:20]: That is an old code style.  Circa the 98 standard, there was a convention to use the base class name to scope calls like this.  Another one that is used in some codes is `this->reset()`.  It also ensures that you are not calling a `reset` function from a derived class.

File: src/meshReaders/Hex_Mesh_Reader.cc (hunk 6 above)
Comment (KineticTheory) [reply src/meshReaders/Hex_Mesh_Reader.cc:233]: You've seem my other PRs; I'm all about fixing spelling errors.  MSVC has a built in spell checker that works like Word.  It underlines the misspelled words.

**Comment by KineticTheory** [comment]
> I have a bit of a hard time parsing the codecov/project results... is it saying this PR causes coverage of CDI to decrease? I'm not sure that makes sense to me.

The number of covered lines is constant, but the total number of lines of code is less (I removed 14 lines of code in this PR), that means that coverage percent decreases.  I usually only worry about covering new lines of code.

File: src/meshReaders/Hex_Mesh_Reader.hh (hunk 5 above)
Comment (KineticTheory) [reply src/meshReaders/Hex_Mesh_Reader.hh]: I just pushed a fix for this.  Again - old code.

**Comment by KineticTheory** [comment]
d0e0e9f only changed comment blocks.  I'm not going to wait for the tests to run.  Merging now.

File: src/meshReaders/Hex_Mesh_Reader.hh (hunk 5 above)
Comment (keadyk) [reply src/meshReaders/Hex_Mesh_Reader.hh]: (LOL, should've realized it would treat that as actual HTML :D )

**Pull request merged**